Eastern-Research-Group/mywaterway
# Feature/215 cypress updates

## Related Issues:
* https://jira.epa.gov/browse/HMW-215

## Main Changes:
* Fixed broken tests. Some were due to data changes on the web services side and some were due to changes we made.
* Switched some tests from using `findByText` to `findByRole`. 
* Updated commands to match Clean School Bus. Now `npm run coverage` will run cypress in headless mode, which is faster and more reliable. 
* Made the snapshots smaller so the snapshot image sizes are the same when running cypress with the ui verses running cypress in headless mode.
* Added a config variable, so that `cypress-image-snapshot` will always pass the image comparison tests. This was done because the image comparison tests are really unreliable (i.e., test would fail even though there is no discernable difference in the snapshots). So when running the visual tests we will have to manually review the diffs in the `__diff_output__` folders.
  * For the next release cycle I want to investigate other packages for the image comparison tests, because this one hasn't been updated in a long time and is supposed to work against cypress 4.5. I think [cypress-image-diff](https://github.com/uktrade/cypress-image-diff) may be a suitable alternative, but due to the time crunch I decided to punt this. 
* Upgraded cypress to version 10.
  * This is a really big update. Cypress completely changed the UI, added component level testing, changed the config file structure, and deprecated the plugins file. 
* Converted the cypress code over to using TypeScript. 
* Split up the community page tests, since these still cause cypress v10 to run out of memory.

## Steps To Test:
1. Run `npm run cypress`
2. Run a few of the tests and make sure they pass. No need to run them all, as they will take 30+ minutes to run and I've already tested them all. 
3. In the `app/package.json` file change line 37 
    * FROM `"coverage": "concurrently -kc \"blue.dim,green.dim,yellow.dim\" -n server,client,cypress \"npm:server\" \"npm:client\" \"cypress run --browser edge\""`
    * TO `"coverage": "concurrently -kc \"blue.dim,green.dim,yellow.dim\" -n server,client,cypress \"npm:server\" \"npm:client\" \"cypress run --spec cypress/e2e/about.cy.ts --browser edge\""`
    * NOTE:  This is just to shorten the run time, as it will take 30+ minutes to run all of the specs 
5. Run `npm run coverage`.
6. Open the code coverage report located in `app/coverage/lcov-report/index.html`
7. Verify the report has data

## TODO
* [ ] - Look into better ways of handling the visual tests. Possibly by switching to the [cypress-image-diff](https://github.com/uktrade/cypress-image-diff) package.
* [ ] - Add more tests for the individual community tabs, since currently we only have tab level spec files for `Identified Issues`, `Monitoring`, and `Protect`. This may just involve pulling some of the tests out of the `community.cy.ts` file. 
* [ ] - Add some component level tests. I think we should just start by adding component level tests to new components that we create. Later on, after we complete some of these bigger tasks, we can go back and create component level tests for the existing components.
* [ ] - Continue switching to `findByRole` in most places.
